xarray/core/rolling_exp.py

```diff
@@ -128,9 +129,41 @@ def mean(self, keep_attrs: bool | None = None) -> T_DataWithCoords:
         if keep_attrs is None:
             keep_attrs = _get_keep_attrs(default=True)
 
-        return self.obj.reduce(
-            move_exp_nanmean, dim=self.dim, alpha=self.alpha, keep_attrs=keep_attrs
-        )
+        dim_order = self.obj.dims
+
+        from xarray.core.dataset import Dataset
+
+        if isinstance(self.obj, Dataset):
+            new_ds = self.obj.copy()
+            if not keep_attrs:
+                new_ds.attrs = {}
+            for name, var in self.obj.variables.items():
+                if self.dim in self.obj[name].dims:
+                    new_ds[name] = apply_ufunc(
+                        move_exp_nanmean,
+                        var,
+                        input_core_dims=[[self.dim]],
+                        kwargs=dict(alpha=self.alpha, axis=-1),
+                        output_core_dims=[[self.dim]],
+                        exclude_dims={self.dim},
+                        keep_attrs=keep_attrs,
+                    )
+                # Currently we test for attrs being removed on all arrays, even those
+                # without the dim, not sure if we should be doing this or not (we can
+                # indent a line and then only remove attrs on arrays we compute on)
+                if not keep_attrs:
+                    new_ds[name].attrs = {}
+            return new_ds.transpose(*dim_order)
+        else:
+            return apply_ufunc(
+                move_exp_nanmean,
+                self.obj,
+                input_core_dims=[[self.dim]],
+                kwargs=dict(alpha=self.alpha, axis=-1),
+                output_core_dims=[[self.dim]],
+                exclude_dims={self.dim},
+                keep_attrs=keep_attrs,
```

Comment:
This is basically the nice shortened version that I think we should try and get to if we use `apply_ufunc` over `reduce`...